File: Python/Encrypt credentials/Encryption sample/services/updatecredentialsservice.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT license.

import requests
from flask import json
from models.credentialsdetails import CredentialsDetails
from models.credentialsdetailsrequest import CredentialsDetailsRequest
from services.asymmetrickeyencryptor import AsymmetricKeyEncryptor
from utils import Utils


class UpdateCredentialsService:

    headers = None

    def update_datasource(self, access_token, cred_type, privacy_level, credentials_array, gateway, datasource_id):
        ''' Updates data source with encrypted credentials

        Args:
            access_token (str): Access token to call API
            cred_type (str): Type of the credentials (i.e. Basic, Windows, OAuth2)
            privacy_level (str): Privacy level
            credentials_array (dict): Credentials based on the user input of the credentials type
            gateway (Gateway): Gateway response
            datasource_id (str): Data source id to encrpt

        Returns:
            Response: Response from the API call
        '''
        gateway_id = gateway['id']
        public_key = gateway['publicKey']

        # Serialize credentials for encryption
        serialized_credentials = Utils.serialize_credentials(credentials_array, cred_type)

        # On-premises gateway contains name property
        # Use on-premises gateway
        if 'name' in gateway:
            # Encrypt the credentials Asymmetric Key Encryption
            asymmetric_encryptor_service = AsymmetricKeyEncryptor(public_key)
            encrypted_credentials_string = asymmetric_encryptor_service.encode_credentials(serialized_credentials)
            encrypted_data = encrypted_credentials_string
            encrypted_connection = 'Encrypted'

        # Use cloud gateway
        else:
            encrypted_data = serialized_credentials
            encrypted_connection = 'NotEncrypted'

        # Credential Details class object for request body
        credentials_details = CredentialsDetails(cred_type, encrypted_data, encrypted_connection, privacy_level)

        # Converting CredentialDetails class object to json string
        credentials_details_req = CredentialsDetailsRequest(credentials_details.__dict__)

        return self.make_update_datasource_patch_request(credentials_details_req, gateway_id, datasource_id, access_token)

    def make_update_datasource_patch_request(self, credentials_details_req, gateway_id, datasource_id, access_token):
        ''' Makes API call

        Args:
            credentials_details_req (CredentialsDetailsRequest): Credentials update request body
            gateway_id (str): Gateway Id
            datasource_id (str): Data source id to encrpt
            access_token (str): Access token to call API

        Returns:
            Response: Response from the API call
        '''

        self.headers = {'Content-Type': 'application/json', 'Authorization': 'Bearer ' + access_token}

        # Gateways - Update Datasource Power BI REST API
        # https://docs.microsoft.com/en-us/rest/api/power-bi/gateways/updatedatasource
        endpoint_url = f'https://api.powerbi.com/v1.0/myorg/gateways/{gateway_id}/datasources/{datasource_id}'

        api_response = requests.patch(endpoint_url, data=json.dumps(credentials_details_req.__dict__), headers=self.headers)

        return api_response

File: Java/EncryptCredentials/src/main/java/com/encryptcredentialsample/encryptcredential/services/UpdateCredentialsService.java
// ----------------------------------------------------------------------------
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
// ----------------------------------------------------------------------------

package com.encryptcredentialsample.encryptcredential.services;

import java.io.IOException;
import org.apache.http.client.ClientProtocolException;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.http.client.HttpComponentsClientHttpRequestFactory;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.client.RestTemplate;

import com.encryptcredentialsample.encryptcredential.config.Config;
import com.encryptcredentialsample.encryptcredential.models.CredentialDetails;
import com.encryptcredentialsample.encryptcredential.models.CredentialDetailsRequestBody;
import com.encryptcredentialsample.encryptcredential.models.Gateway;

public class UpdateCredentialsService {

	// Encrypts and makes PATCH request to update datasource
	public static ResponseEntity<String> updateDatasource(
			String accessToken, 
			String credType,
			String privacyLevel, 
			String[] credentialsArray, 
			String gatewayId, 
			String datasourceId) throws Exception {

		// Serialize credentials for encryption
		String serializedCredentials = Utils.serializeCredentials(credentialsArray, credType);

		String encryptedCredentialsString = null;
		String encryptedConnection = null;

		Gateway gateway = new Gateway();
		try {
			gateway = GetDatasourceData.getGateway(accessToken, gatewayId);
		}
		catch (HttpClientErrorException e) {
			if (!e.getStatusText().toString().equals("Not Found")) {
				throw e;
			}
		}
		
		// On-premises gateway contains name property
        // Use on-premises gateway
        if (gateway.name != null) {
        	encryptedConnection = "Encrypted";
    		// Encrypt the credentials Asymmetric Key Encryption
    		AsymmetricKeyEncryptorService credentialsEncryptor = new AsymmetricKeyEncryptorService(gateway.publicKey);
        	encryptedCredentialsString = credentialsEncryptor.encodeCredentials(serializedCredentials);
        } else {
			// Use cloud gateway
        	encryptedConnection = "NotEncrypted";
        	encryptedCredentialsString = serializedCredentials;	
        }

        // Credential Details class object for request body
        CredentialDetails credentialDetails = new CredentialDetails(credType, encryptedCredentialsString, encryptedConnection, privacyLevel); 

		// Converting CredentialDetails class object to json string
		CredentialDetailsRequestBody requestBodyObj = new CredentialDetailsRequestBody(credentialDetails);

		return makeUpdateDataSourcePatchRequest(requestBodyObj, gatewayId, datasourceId, accessToken);
	}

	// Make an HTTP Patch request to update the encrypted credentials of the data source
	public static ResponseEntity<String> makeUpdateDataSourcePatchRequest(
			CredentialDetailsRequestBody requestBody, 
			String gatewayId, 
			String datasourceId, 
			String accessToken) throws ClientProtocolException, IOException {
		
		// Gateways - Update Datasource Power BI REST API
		// https://docs.microsoft.com/en-us/rest/api/power-bi/gateways/updatedatasource
		String endPointUrl = Config.powerBiApiUrl + "v1.0/myorg/gateways/" + gatewayId + "/datasources/" + datasourceId;

		// Request header
		HttpHeaders reqHeader = Utils.generateAuthorizationHeaders(accessToken);
		
		// HTTP entity object - contains header and body
		HttpEntity<CredentialDetailsRequestBody> reqEntity = new HttpEntity<CredentialDetailsRequestBody> (requestBody, reqHeader);

		// Call REST API
		RestTemplate updateDatasource = new RestTemplate(new HttpComponentsClientHttpRequestFactory());
		return updateDatasource.exchange(endPointUrl, HttpMethod.PATCH, reqEntity, String.class);
	}
}
